rotkehlchen/chain/aggregator.py

```diff
@@ -1286,72 +1287,156 @@ def get_evm_manager(
 
         Returns a list of tuples of the address and the chain it was added in.
         """
-        filtered_accounts = self.filter_active_evm_addresses(accounts)
-        added_accounts = []
-        for chain, account in filtered_accounts:
-            try:
-                self.modify_blockchain_accounts(
-                    blockchain=chain,
-                    accounts=[account],
-                    append_or_remove='append',
+        added_accounts: list[tuple[SUPPORTED_EVM_CHAINS, ChecksumEvmAddress]] = []
+        # Distinguish between contracts and EOAs
+        for account in accounts:
+            if self.is_contract(account, SupportedBlockchain.ETHEREUM):
+                added_chains = self.track_evm_address(account, [SupportedBlockchain.ETHEREUM])
+                if len(added_chains) == 1:  # Is always either 1 or 0 since is only for ethereum
+                    added_accounts.append((SupportedBlockchain.ETHEREUM, account))
+            else:
+                chains_to_check = []
+                for chain in typing.get_args(SUPPORTED_EVM_CHAINS):
+                    if account not in self.accounts.get(chain):
+                        chains_to_check.append(chain)
+
+                added_accounts += self.check_chains_and_add_accounts(
+                    account=account,
+                    chains=chains_to_check,
                 )
-            except InputError:
-                log.debug(f'Not adding {account} to {chain} since it already exists')
+
+        return added_accounts
+
+    def detect_evm_accounts(
+            self,
+            progress_handler: Optional['ProgressUpdater'] = None,
+    ) -> list[tuple[SUPPORTED_EVM_CHAINS, ChecksumEvmAddress]]:
+        """
+        Detects user's EVM accounts on different chains and adds them to the tracked accounts.
+        1. Iterates through already added addresses
+        2. Checks whether the address is the chain that it is currently added to
+        3. If it is not, checks whether the address is active in a chain
+        4. If it is active, adds it to the tracked accounts on that chain
+
+        Returns a list of tuples of (chain, address) for the freshly detected accounts.
+        """
+        current_accounts: dict[ChecksumEvmAddress, list[SUPPORTED_EVM_CHAINS]] = defaultdict(list)
+        chain: SUPPORTED_EVM_CHAINS
+        for chain in typing.get_args(SUPPORTED_EVM_CHAINS):
+            chain_accounts = self.accounts.get(chain)
+            for account in chain_accounts:
+                current_accounts[account].append(chain)
+
+        all_evm_chains = set(typing.get_args(SUPPORTED_EVM_CHAINS))
+        added_accounts: list[tuple[SUPPORTED_EVM_CHAINS, ChecksumEvmAddress]] = []
+        for account, chains in current_accounts.items():
+            if progress_handler is not None:
+                progress_handler.new_step(f'Checking {account} EVM chain activity')
+
+            if self.is_contract(account, chains[0]):
```

Comment:
Will also add it as a comment, but no, not assuming chain[0] being ETH.

I rather assume that if an address is a contract in one chain, it will be a contract in all other chains as well.

And thus checking only the first chain.